lib/binance/futures/rest/http_client.ex

```diff
@@ -0,0 +1,160 @@
+defmodule Binance.Futures.Rest.HTTPClient do
+  @endpoint "https://fapi.binance.com"
+
+  def get_binance(url, headers \\ []) do
+    HTTPoison.get("#{@endpoint}#{url}", headers)
+    |> parse_response
+  end
+
+  def delete_binance(url, headers \\ []) do
+    HTTPoison.delete("#{@endpoint}#{url}", headers)
+    |> parse_response
+  end
+
+  def get_binance(url, params, secret_key, api_key) do
+    case prepare_request(url, params, secret_key, api_key) do
+      {:error, _} = error ->
+        error
+
+      {:ok, url, headers} ->
+        get_binance(url, headers)
+    end
+  end
+
+  def delete_binance(url, params, secret_key, api_key) do
+    case prepare_request(url, params, secret_key, api_key) do
+      {:error, _} = error ->
+        error
+
+      {:ok, url, headers} ->
+        delete_binance(url, headers)
+    end
+  end
+
+  defp prepare_request(url, params, secret_key, api_key) do
+    case validate_credentials(secret_key, api_key) do
+      {:error, _} = error ->
+        error
+
+      _ ->
+        headers = [{"X-MBX-APIKEY", api_key}]
+        receive_window = 5000
+        ts = DateTime.utc_now() |> DateTime.to_unix(:millisecond)
+
+        params =
+          Map.merge(params, %{
+            timestamp: ts,
+            recvWindow: receive_window
+          })
+
+        argument_string = URI.encode_query(params)
+
+        signature =
+          :crypto.hmac(
+            :sha256,
+            secret_key,
+            argument_string
+          )
+          |> Base.encode16()
+
+        {:ok, "#{url}?#{argument_string}&signature=#{signature}", headers}
+    end
+  end
+
+  def post_binance(url, params) do
+    argument_string =
+      params
+      |> Map.to_list()
+      |> Enum.map(fn x -> Tuple.to_list(x) |> Enum.join("=") end)
+      |> Enum.join("&")
+
+    # generate signature
+    signature =
+      :crypto.hmac(
+        :sha256,
+        Application.get_env(:binance, :secret_key),
+        argument_string
+      )
+      |> Base.encode16()
+
+    body = "#{argument_string}&signature=#{signature}"
+
+    case HTTPoison.post("#{@endpoint}#{url}", body, [
+           {"X-MBX-APIKEY", Application.get_env(:binance, :api_key)}
+         ]) do
+      {:error, err} ->
+        {:error, {:http_error, err}}
+
+      {:ok, response} ->
+        case Poison.decode(response.body) do
+          {:ok, data} -> {:ok, data}
+          {:error, err} -> {:error, {:poison_decode_error, err}}
+        end
+    end
+  end
+
+  def put_binance(url, params) do
+    argument_string =
+      params
+      |> Map.to_list()
+      |> Enum.map(fn x -> Tuple.to_list(x) |> Enum.join("=") end)
+      |> Enum.join("&")
```

Comment:
This parsing part duplicate in all of the functions. Can we extract that out.

Reply:
👌